### PR TITLE
Improve docs for Box3 and BoxHelper

### PR DESCRIPTION
**Pull request by jotaro-sama**
The `BoxHelper` class has really vague docs that don't really make it clear what the workflow for using it should be, or what its use cases are. I updated its documentation page and the "example" section in particular so that it's more useful for developers.

### CODE REVIEW

File: docs/api/en/helpers/BoxHelper.html [hunk 1]
```diff
@@ -13,8 +13,8 @@
 		<h1>[name]</h1>
 
 		<p class="desc">
-			Helper object to show the world-axis-aligned bounding box around an object.
-
+			Helper object to show the world-axis-aligned bounding box around an object. 
+			It can be used in synergy with [page:Box3] items in order to easily compute the [link:https://en.wikipedia.org/wiki/Minimum_bounding_box Minimum Bounding Boxes] for objects, and use them for fast collision detection. See [page:Box3.intersect].
```
Comment (gkjohnson) [inline docs/api/en/helpers/BoxHelper.html]: I feel this is misleading. You don't need a BoxHelper object to create bounding boxes for objects and I don't think it should be implied that the box helper should be used for anything other than debugging.

Comment (jotaro-sama) [reply docs/api/en/helpers/BoxHelper.html]: You don't need it, but AFAIK using a BoxHelper is the only way to automatically resize the bounding box every time the object changes position, rotation or scale. I'd argue its visual use for debug is actually secondary in relevance compared to what the average developer can get from it as an easy way to keep your bounding boxes consistent. If there's another way I'm not aware of I'll be glad to learn about it. I can rephrase it to say it's used to compute an axis-aligned bounding box though.

Comment (jotaro-sama) [reply docs/api/en/helpers/BoxHelper.html]: While you don't really need it, it would be nice to at least give some hint that you can also calculate the bounding box this way while debugging. Also, I just edited the description to point out that it's the axis-aligned bounding box to be calculated and not the minimum one, but if that is true shouldn't the Box3 page of the docs also be corrected? I can take care of it as well if that's the case.

> Represents a box or cube in 3D space. The main purpose of this is to represent the Minimum Bounding Boxes for objects. 
https://threejs.org/docs/#api/en/math/Box3

Comment (gkjohnson) [reply docs/api/en/helpers/BoxHelper.html]: I think the docs should captured the designed intent of the class, which is just for debugging -- I don't think it's a good idea to encourage this kind of use of BoxHelper in the docs. The BoxHelper inherits from `Object3D` so while I'm not sure if it makes sense to do it implies that this will work:

```js
var box3 = new THREE.Box3();
box3.setFromObject( helper );
```

But for any production use the box3 object should just call `setFromObject` on the original target.

> if that is true isn't the Box3 page of the docs also incorrect?

I would think so. To me a "minimum bounding box" needs to be able to be oriented to minimize the volume of the box. "Minimum axis-aligned bounding box" might be better.

Comment (jotaro-sama) [reply docs/api/en/helpers/BoxHelper.html]: > I think the docs should captured the designed intent of the class, which is just for debugging -- I don't think it's a good idea to encourage this kind of use of BoxHelper in the docs.
Got it, I'll try to make it clearer in the description of the class/

> The BoxHelper inherits from `Object3D` so while I'm not sure if it makes sense to do it implies that this will work:
> 
> ```js
> var box3 = new THREE.Box3();
> box3.setFromObject( helper );
> ```
> 
> But for any production use the box3 object should just call `setFromObject` on the original target.

I agree, but wouldn't it be better to at least point out in the example how you can use the `update()` method? I can slightly edit the example so that it both shows the correct way of updating the Box3 in production and the way you update the helper. 

> > if that is true isn't the Box3 page of the docs also incorrect?
> 
> I would think so. To me a "minimum bounding box" needs to be able to be oriented to minimize the volume of the box. "Minimum axis-aligned bounding box" might be better.

Gonna correct that as well then.

File: docs/api/en/helpers/BoxHelper.html [hunk 2]
```diff
@@ -28,10 +28,25 @@ <h2>Example</h2>
...
+		var sphereObject = new THREE.Mesh( sphereGeom, new THREE.MeshBasicMaterial( 0xff0000 ) );
+
+		// Creating the helper and displaying it in the scene
+		var helper = new THREE.BoxHelper( sphereObject, 0xffff00 );
+		scene.add( helper );
+
+		// Creating the actual bounding box (on which we can test intersection with other Box3 objects)
+		var box3 = new THREE.Box3();
+		// Shape and position the box after the object
+		box3.setFromObject( sphere );
```
Comment (Mugen87) [inline docs/api/en/helpers/BoxHelper.html]: I think it should be `sphereObject` and not `sphere`.

File: docs/api/en/helpers/BoxHelper.html [hunk 3]
```diff
@@ -28,10 +28,25 @@ <h2>Example</h2>
...
+
+		// Let's move the sphere
+		sphereObject.position.set( 7, 7, 7 );
+
+		// Whenever the object is moved or rotated or scaled, we update both the bounding box and the helper
+		helper.update(); 
+		box3.setFromObject( sphereObject );
```
Comment (Mugen87) [inline docs/api/en/helpers/BoxHelper.html]: Usually you do not want generate a new bounding box each time the object is updated. Instead, you would copy the bounding box from the geometry and just apply the world matrix of the object.
```js
sphereGeom.computeBoundingBox(); // just do this once

// in your animation loop

box.copy( object.geometry.boundingBox ).applyMatrix4( object.worldMatrix );
```
You would do the same with bounding sphere.

File: docs/api/en/math/Box3.html [hunk 4]
```diff
@@ -12,8 +12,7 @@ <h1>[name]</h1>
 
 		<p class="desc">
 			Represents a box or cube in 3D space. The main purpose of this is to represent
-			the [link:https://en.wikipedia.org/wiki/Minimum_bounding_box Minimum Bounding Boxes]
-			for objects.
+			the world-axis-aligned bounding boxes for objects, which are used for fast collision detection.
```
Comment (Mugen87) [inline docs/api/en/math/Box3.html]: AABBs are not only used for collision detection. I would remove this clause.

File: docs/api/en/helpers/BoxHelper.html (hunk 2 above)
Comment (jotaro-sama) [reply docs/api/en/helpers/BoxHelper.html]: Yeah, my bad. It was a typo.

File: docs/api/en/math/Box3.html (hunk 4 above)
Comment (jotaro-sama) [reply docs/api/en/math/Box3.html]: Got it, gonna remove this statement.

File: docs/api/en/helpers/BoxHelper.html (hunk 3 above)
Comment (jotaro-sama) [reply docs/api/en/helpers/BoxHelper.html]: Didn't know `setFromObject` had the effect of creating a new bounding box from scratch instead of copying the relevant parameters only. Thanks for the insight. I'm gonna change the example as you suggest, but may I ask why don't we change `setFromObject` to do those operations instead of re-creating a new Box3 object every time? Are there cases when that is the intended behavior?

Also, is the World matrix what is commonly referred to as a Model matrix (mapping from model coordinates to world coordinates, containing all the transformations)?

Comment (Mugen87) [reply docs/api/en/helpers/BoxHelper.html]: >  but may I ask why don't we change setFromObject to do those operations instead of re-creating a new Box3 object every time?

`Box3.setFromObject()` should compute an AABB from scratch by evaluating geometry data. I referred in my previous post to the case where you apply a transformation matrix to an _existing_ AABB. These are two different operations and both are needed in different scenarios.

> Also, is the World matrix what is commonly referred to as a Model matrix

Yep.

Comment (jotaro-sama) [reply docs/api/en/helpers/BoxHelper.html]: Understood, thanks for the clarification. I'm gonna do all of the requested changes.

**Comment by jotaro-sama** [comment]
@Mugen87 I made the changes you requested.

File: docs/api/en/helpers/BoxHelper.html [hunk 5]
```diff
@@ -28,10 +28,32 @@ <h2>Example</h2>
 
 
 		<code>
-		var sphere = new THREE.SphereGeometry();
-		var object = new THREE.Mesh( sphere, new THREE.MeshBasicMaterial( 0xff0000 ) );
-		var box = new THREE.BoxHelper( object, 0xffff00 );
-		scene.add( box );
+		// Creating the object whose bounding box we want to compute
```
Comment (Mugen87) [inline docs/api/en/helpers/BoxHelper.html]: Sorry but I think the previous code example was totally sufficient. `BoxHelper` is just a helper class and the example code should demonstrate its creation. The actual handling of a `Box3` does not belong to this page.

Comment (jotaro-sama) [reply docs/api/en/helpers/BoxHelper.html]: Yeah it makes sense, I guess I should rather add a decent example to Box3 as it doesn't have one and its workflow isn't that immediate.

Comment (Mugen87) [reply docs/api/en/helpers/BoxHelper.html]: Correct.

File: docs/api/en/math/Box3.html [hunk 6]
```diff
@@ -12,11 +12,37 @@ <h1>[name]</h1>
...
+			sphereGeom, 
+			new THREE.MeshBasicMaterial( 0xff0000 ) 
+		);
+
+		// Creating the actual bounding box with Box3
+		sphereObject.geometry.computeBoundingBox();
+		var BBox = new THREE.Box3(
+			sphereObject.geometry.boundingBox.min,
+  			sphereObject.geometry.boundingBox.max
+		);
```
Comment (Mugen87) [inline docs/api/en/math/Box3.html]: I don't understand why you pass in ctor paramters when you use the object for a copy operation two lines later. Besides `BBox` is no good variable name. `box` should be sufficient.

File: docs/api/en/math/Box3.html [hunk 7]
```diff
@@ -12,11 +12,37 @@ <h1>[name]</h1>
...
+
+		// Moving the sphere
+		sphereObject.position.set( 7, 7, 7 );
+
+		// Updating the bounding box
+		// (to repeat each time the object is moved, rotated or scaled) 
+		BBox.copy( sphereObject.geometry.boundingBox ).applyMatrix4( sphereObject.matrixWorld );
```
Comment (Mugen87) [inline docs/api/en/math/Box3.html]: This code is wrong since you did not update the world matrix after setting the position.

**Comment by Mugen87** [comment]
Please don't take this personally but such code sections should be written by devs who are more familiar with the library. Otherwise it take the reviewers a lot of time to correct this stuff.

File: docs/api/en/math/Box3.html (hunk 6 above)
Comment (jotaro-sama) [reply docs/api/en/math/Box3.html]: Because the `copy` was supposed to be placed inside the animation loop. I'm editing the samples so that it is clear which part is the initialization and which is the updating. However, I think using clone makes the code sample more elegant and uniform so I'm gonna change it.

File: docs/api/en/math/Box3.html (hunk 7 above)
Comment (jotaro-sama) [reply docs/api/en/math/Box3.html]: Oh sorry, for some reason my tests worked even without updating it so I forgot it. I'm gonna fix it asap.

**Comment by jotaro-sama** [comment]
Fixes are there, I hope everything is fine now.

**Comment by jotaro-sama** [comment]
Fixes have been committed, I hope everything is fine now.

**Comment by jotaro-sama** [comment]
Fixes have been committed, everything should be fine now.

**Comment by mrdoob** [comment]
Thanks!
